core/src/main/java/tc/oc/pgm/channels/MessageChannel.java

```diff
@@ -0,0 +1,292 @@
+import static net.kyori.adventure.text.Component.text;
+import static net.kyori.adventure.text.Component.translatable;
+import static tc.oc.pgm.util.text.TextException.*;
+
+import cloud.commandframework.arguments.parser.ParserParameters;
+import cloud.commandframework.arguments.standard.StringArgument;
+import cloud.commandframework.paper.PaperCommandManager;
+import io.leangen.geantyref.TypeToken;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import net.kyori.adventure.sound.Sound;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextDecoration;
+import org.bukkit.Bukkit;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.Nullable;
+import tc.oc.pgm.api.PGM;
+import tc.oc.pgm.api.Permissions;
+import tc.oc.pgm.api.channels.Channel;
+import tc.oc.pgm.api.event.ChannelMessageEvent;
+import tc.oc.pgm.api.integration.Integration;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.api.setting.SettingKey;
+import tc.oc.pgm.api.setting.SettingValue;
+import tc.oc.pgm.util.MessageSenderIdentity;
+import tc.oc.pgm.util.Players;
+import tc.oc.pgm.util.bukkit.OnlinePlayerMapAdapter;
+import tc.oc.pgm.util.named.NameStyle;
+
+public class MessageChannel implements Channel<MatchPlayer> {
+
+  private final ChannelManager manager;
+  private final OnlinePlayerMapAdapter<MessageSenderIdentity> setChannel, lastMessagedBy;
+
+  private static final Sound SOUND = sound(key("random.orb"), Sound.Source.MASTER, 1f, 1.2f);
+
+  public MessageChannel(ChannelManager manager) {
+    this.manager = manager;
+    this.setChannel = new OnlinePlayerMapAdapter<MessageSenderIdentity>(PGM.get());
+    this.lastMessagedBy = new OnlinePlayerMapAdapter<MessageSenderIdentity>(PGM.get());
+  }
+
+  @Override
+  public String getDisplayName() {
+    return "messages";
+  }
+
+  @Override
+  public String[] getAliases() {
+    return new String[] {"msg", "tell", "r"};
+  }
+
+  @Override
+  public Character getShortcut() {
+    return '@';
+  }
+
+  @Override
+  public MatchPlayer getTarget(MatchPlayer sender, Map<String, ?> arguments) {
+    MatchPlayer target = (MatchPlayer) arguments.get("target");
+    if (target == null) {
+      return getSelectedTarget(sender);
+    } else {
+      checkSettings(target, sender);
+
+      if (!target.equals(getSelectedTarget(sender))
+          && !(manager.getSelectedChannel(sender) instanceof MessageChannel)) {
+        this.lastMessagedBy.put(
+            sender.getBukkit(), new MessageSenderIdentity(sender.getBukkit(), target.getBukkit()));
+      }
+
+      if (!sender.equals(getSelectedTarget(target))
+          && !(manager.getSelectedChannel(target) instanceof MessageChannel)) {
+        this.lastMessagedBy.put(
+            target.getBukkit(), new MessageSenderIdentity(target.getBukkit(), sender.getBukkit()));
+      }
```

Comment:
this should probably update as part of post-sending hooks (ie: same place ping sounds are placed at)